In [1]:
import pandas as pd
import numpy as np

In [93]:
df = pd.read_csv('Mercedes-Benz.csv')
df

ID       y  X0 X1  X2 X3 X4  X5 X6 X8  ...  X375  X376  X377  X378  \
0        0  130.81   k  v  at  a  d   u  j  o  ...     0     0     1     0   
1        6   88.53   k  t  av  e  d   y  l  o  ...     1     0     0     0   
2        7   76.26  az  w   n  c  d   x  j  x  ...     0     0     0     0   
3        9   80.62  az  t   n  f  d   x  l  e  ...     0     0     0     0   
4       13   78.02  az  v   n  f  d   h  d  n  ...     0     0     0     0   
...    ...     ...  .. ..  .. .. ..  .. .. ..  ...   ...   ...   ...   ...   
4204  8405  107.39  ak  s  as  c  d  aa  d  q  ...     1     0     0     0   
4205  8406  108.77   j  o   t  d  d  aa  h  h  ...     0     1     0     0   
4206  8412  109.22  ak  v   r  a  d  aa  g  e  ...     0     0     1     0   
4207  8415   87.48  al  r   e  f  d  aa  l  u  ...     0     0     0     0   
4208  8417  110.85   z  r  ae  c  d  aa  g  w  ...     1     0     0     0   

      X379  X380  X382  X383  X384  X385  
0        0     0     0     0     0     0  
1        0     0     0     0     0     0  
2        0     0     1     0     0     0  
3        0     0     0     0     0     0  
4        0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...  
4204     0     0     0     0     0     0  
4205     0     0     0     0     0     0  
4206     0     0     0     0     0     0  
4207     0     0     0     0     0     0  
4208     0     0     0     0     0     0  

[4209 rows x 378 columns]

In [94]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Columns: 378 entries, ID to X385
dtypes: float64(1), int64(369), object(8)
memory usage: 12.1+ MB


# Checking out of 378 column how many columns have categorical vlaues

In [95]:
cat_col = df.select_dtypes(object)
cat_col.head()

X0 X1  X2 X3 X4 X5 X6 X8
0   k  v  at  a  d  u  j  o
1   k  t  av  e  d  y  l  o
2  az  w   n  c  d  x  j  x
3  az  t   n  f  d  x  l  e
4  az  v   n  f  d  h  d  n

In [96]:
# Printing the unique value in each column

for col in cat_col.columns:
    print(col, ':', len(cat_col[col].unique()), 'labels')

X0 : 47 labels
X1 : 27 labels
X2 : 44 labels
X3 : 7 labels
X4 : 4 labels
X5 : 29 labels
X6 : 12 labels
X8 : 25 labels


In [97]:
#To check how many columns will be created if i perform one end hot encoding.
#dropping the first column
pd.get_dummies(cat_col, drop_first=True).shape 


(4209, 187)

# As the number of features or columns increases it leads to curse of dimensionaltiy. it may impact the accuracy level.

# so we will take top 10 unique value from each column as we have 8 column.

In [98]:
cat_col.X2.value_counts().sort_values(ascending=False).head(10)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
Name: X2, dtype: int64

In [99]:
top_10 = [x for x in cat_col.X2.value_counts().sort_values(ascending=False).head(10).index]

In [100]:
top_10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [101]:
#Creating 10 binary columns

for label in top_10:
    cat_col[label] = np.where(cat_col['X2']==label,1,0)

cat_col[['X2']+top_10].head()

<ipython-input-101-8c1395894886>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cat_col[label] = np.where(cat_col['X2']==label,1,0)


X2  as  ae  ai  m  ak  r  n  s  f  e
0  at   0   0   0  0   0  0  0  0  0  0
1  av   0   0   0  0   0  0  0  0  0  0
2   n   0   0   0  0   0  0  1  0  0  0
3   n   0   0   0  0   0  0  1  0  0  0
4   n   0   0   0  0   0  0  1  0  0  0

# Now Creating the function for getting the top 10 variable from each column

In [102]:
def one_hot_encoding(df, columns, top_label):
    for label in top_label:
        df[columns+'_'+label] = np.where(df[columns]==label, 1,0)
        


In [108]:
df = pd.read_csv('Mercedes-Benz.csv')

#taking the catogorical column
df1 = df.select_dtypes(object)
df1.head()

X0 X1  X2 X3 X4 X5 X6 X8
0   k  v  at  a  d  u  j  o
1   k  t  av  e  d  y  l  o
2  az  w   n  c  d  x  j  x
3  az  t   n  f  d  x  l  e
4  az  v   n  f  d  h  d  n

In [109]:
x0_top_10 = [x for x in df1.X0.value_counts().sort_values(ascending=False).head(10).index]
x1_top_10 = [x for x in df1.X1.value_counts().sort_values(ascending=False).head(10).index]
x2_top_10 = [x for x in df1.X2.value_counts().sort_values(ascending=False).head(10).index]
x3_top_10 = [x for x in df1.X3.value_counts().sort_values(ascending=False).head(10).index]
x4_top_10 = [x for x in df1.X4.value_counts().sort_values(ascending=False).head(10).index]
x5_top_10 = [x for x in df1.X5.value_counts().sort_values(ascending=False).head(10).index]
x6_top_10 = [x for x in df1.X6.value_counts().sort_values(ascending=False).head(10).index]
x8_top_10 = [x for x in df1.X8.value_counts().sort_values(ascending=False).head(10).index]

In [110]:
one_hot_encoding(df1,'X0', x0_top_10)
one_hot_encoding(df1,'X1', x1_top_10)
one_hot_encoding(df1,'X2', x2_top_10)
one_hot_encoding(df1,'X3', x3_top_10)
one_hot_encoding(df1,'X4', x4_top_10)
one_hot_encoding(df1,'X5', x5_top_10)
one_hot_encoding(df1,'X6', x6_top_10)
one_hot_encoding(df1,'X8', x8_top_10)

<ipython-input-102-668578594232>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[columns+'_'+label] = np.where(df[columns]==label, 1,0)


In [111]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Data columns (total 79 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   X0      4209 non-null   object
 1   X1      4209 non-null   object
 2   X2      4209 non-null   object
 3   X3      4209 non-null   object
 4   X4      4209 non-null   object
 5   X5      4209 non-null   object
 6   X6      4209 non-null   object
 7   X8      4209 non-null   object
 8   X0_z    4209 non-null   int32 
 9   X0_ak   4209 non-null   int32 
 10  X0_y    4209 non-null   int32 
 11  X0_ay   4209 non-null   int32 
 12  X0_t    4209 non-null   int32 
 13  X0_x    4209 non-null   int32 
 14  X0_o    4209 non-null   int32 
 15  X0_f    4209 non-null   int32 
 16  X0_n    4209 non-null   int32 
 17  X0_w    4209 non-null   int32 
 18  X1_aa   4209 non-null   int32 
 19  X1_s    4209 non-null   int32 
 20  X1_b    4209 non-null   int32 
 21  X1_l    4209 non-null   int32 
 22  X1_v    4209 non-null   

# Advantages of  One Hot Encoding of Top Variable
- straight forward to implement
- does not require hrs of variable exploration.
- Does not expand massivley the feature space(number of column in dataset.)

# Disadvantage of One Hot Encoding 
- does not add any information that make the varibale more predictve
- does not keep the information of the ignored label.
